python/hail/typecheck.py

```diff
@@ -0,0 +1,123 @@
+from decorator import decorator, getargspec
+from types import ClassType
+import re
+
+
+def extract(t):
+    m = re.match("<(type|class) '(.*)'>", str(t))
+    if m:
+        return m.groups()[1]
+    else:
+        return str(t)
+
+
+class TypeChecker(object):
+    def __init__(self, f, supported):
+        self.f = f
```

Comment:
I think this would be clearer if you had an interface with two functions: check (your f) and pretty/toString builds the message recursively.